### PR TITLE
Input rework

### PR DESCRIPTION
**Pull request by vitvakatu**

### CODE REVIEW

**Comment by alteous** [comment]
What is the meaning of `win.input.button(three::KEY_ESCAPE).released` now? I take 'released' to mean 'the key was down and now it is up'.

**Comment by vitvakatu** [comment]
Yep, it was one of concerns. It's the opposite of `pressed`, so I can remove it and see `!input.button(ESCAPE).pressed` instead. Is it ok?

**Comment by alteous** [comment]
That would be more correct.

File: examples/anim.rs [hunk 1]
```diff
@@ -134,7 +134,7 @@ fn main() {
     let blend_time = 0.2f32;
     let mut timer = win.input.time();
 
-    while win.update() && !three::KEY_ESCAPE.is_hit(&win.input) {
+    while win.update() && !win.input.button(three::KEY_ESCAPE).pressed {
```
Comment (kvark) [inline examples/anim.rs]: hmm, that worries me. The new code is more verbose, while we need to make it shorter

Comment (vitvakatu) [reply examples/anim.rs]: From my point of view, code is little more verbose but much clearer now. If we're talking about amount of symbols, I don't see any way to reduce it atm.

File: src/input/mod.rs [hunk 2]
```diff
@@ -26,7 +26,7 @@ struct State {
     mouse_pos_ndc: mint::Point2<f32>,
 }
 
-struct Delta {
+struct Diff {
```
Comment (alteous) [inline src/input/mod.rs]: We've gone back to diff?

File: src/input/mod.rs [hunk 3]
```diff
@@ -235,128 +235,148 @@ impl Input {
...
+}
+
+impl Hit for axis::Raw {
+    fn hit(&self, input: &Input) -> bool {
+        input.delta.axes_raw.iter().filter(|&&(id, _)| id == self.id).count() > 0
+    }
+}
+
+pub trait HitCount {
+    type Output;
```
Comment (alteous) [inline src/input/mod.rs]: Surprised to see an associated type here.

File: src/input/mod.rs (hunk 2 above)
Comment (vitvakatu) [reply src/input/mod.rs]: Actually, we can keep it as `Delta`, but then we need to move/rename `Delta` trait. Not a problem, I'm just too lazy to do it and I wanted to check it asap

File: src/input/mod.rs (hunk 3 above)
Comment (vitvakatu) [reply src/input/mod.rs]: That's because of `axis::Key`

Comment (alteous) [reply src/input/mod.rs]: Oh, never mind... I just saw `type Output = (u8, u8)`.

File: src/input/mod.rs [hunk 4]
```diff
@@ -235,128 +235,148 @@ impl Input {
...
-                Some(-1)
-            } else {
-                None
-            }
-        };
-        let timed_state = state.map(|v| v as f32 * self.delta_time());
-        let mut hits = KeyAxisHits::default();
-        hits.pos = self.delta
+impl HitCount for axis::Key {
+    type Output = (u8, u8);
```
Comment (alteous) [inline src/input/mod.rs]: Nice one. :+1:

File: src/input/mod.rs (hunk 3 above)
Comment (vitvakatu) [reply src/input/mod.rs]: Not a big deal in my opinion

File: src/input/mod.rs [hunk 5]
```diff
@@ -235,128 +235,148 @@ impl Input {
...
-        let delta_hits = if hits.pos + hits.neg == 0 {
+        (pos, neg)
+    }
+}
+
+pub trait Delta {
+    type Output;
+
+    fn delta(&self, input: &Input) -> Self::Output;
+    fn timed_delta(&self, input: &Input) -> Option<TimerDuration>;
```
Comment (alteous) [inline src/input/mod.rs]: I think `timed` is good enough.

Comment (vitvakatu) [reply src/input/mod.rs]: Yep, maybe. Also we can reduce code a bit by providing default impl for this method

Comment (alteous) [reply src/input/mod.rs]: Are these methods independent? If so, perhaps we should split them into `Delta` and `TimedDelta`.

**Comment by kvark** [comment]
bors r-
wait, is this still WIP?

**Comment by vitvakatu** [comment]
Sorry, I think I need to rebase it. Don't worry, I'll merge with you as reviewer later, can't rebase it now.

**Comment by kvark** [comment]
awesome!
bors r+

**Pull request merged**